Framework/Core/src/DataInspector.cxx

```diff
@@ -0,0 +1,365 @@
+    }
+    if (context != nullptr) {
+      zmq_ctx_destroy(context);
+    }
+  }
+
+  void send(const std::string& message)
+  {
+    if (socket != nullptr) {
+      zmq_send(socket, message.c_str(), message.length(), ZMQ_DONTWAIT);
+    }
+  }
+
+ private:
+  void* context{nullptr};
+  void* socket{nullptr};
+};
+
+class stdout_capture
+{
+ public:
+  stdout_capture() : is_capturing(false)
+  {
+    fds[READ] = 0;
+    fds[WRITE] = 0;
+    if (pipe(fds) == -1) {
+      throw std::runtime_error{"pipe() error"};
+    }
+    old_fd = dup(fileno(stdout));
+    if (old_fd == -1) {
+      throw std::runtime_error{"dup() error"};
+    }
+  }
+
+  ~stdout_capture()
+  {
+    if (is_capturing) {
+      end();
+    }
+    if (old_fd > 0) {
+      close(old_fd);
+    }
+    if (fds[READ] > 0) {
+      close(fds[READ]);
+    }
+    if (fds[WRITE] > 0) {
+      close(fds[WRITE]);
+    }
+  }
+
+  void start()
+  {
+    if (is_capturing) {
+      end();
+    }
+    fflush(stdout);
+    dup2(fds[WRITE], fileno(stdout));
+    is_capturing = true;
+  }
+
+  bool end()
+  {
+    if (!is_capturing) {
+      return false;
+    }
+    fflush(stdout);
+    dup2(old_fd, fileno(stdout));
+    out.clear();
+
+    std::string buffer;
+    int size, nbytes;
+
+    size = 1024;
+    buffer.resize(size);
+
+    while ((nbytes = read(fds[READ], &(*buffer.begin()), size)) >= size) {
+      out << buffer;
+    }
+    if (nbytes > 0) {
+      buffer.resize(nbytes);
+      out << buffer;
+    }
+    is_capturing = false;
+    return true;
+  }
+
+  std::string get() const
+  {
+    return out.str();
+  }
+
+ private:
+  enum PIPES { READ,
+               WRITE };
+  int fds[2], old_fd;
+  bool is_capturing, init;
+  std::stringstream out;
+};
+
+/* Converts the `data` input a string of bytes as two hexadecimal numbers. Adds
+   a 128 offset to deal with negative values. */
+template <typename T>
+static std::string asBytes(const T* data, uint32_t length)
+{
+  if (length == 0) {
+    return "";
+  }
+  std::stringstream buffer;
+  buffer << std::hex << std::setfill('0')
+         << std::setw(2) << static_cast<unsigned>(data[0] + 128);
+  for (uint32_t i = 1; i < length; i++) {
+    buffer << ' ' << std::setw(2) << static_cast<unsigned>(data[i] + 128);
+  }
+  return buffer.str();
+}
+
+/* Replaces (in place) all occurences of `pattern` in `data` with `substitute`.
+   `substitute` may be wider than `pattern`. */
+static void replaceAll(
+  std::string& data,
+  const std::string& pattern,
+  const std::string& substitute)
+{
+  std::string::size_type pos = data.find(pattern);
+  while (pos != std::string::npos) {
+    data.replace(pos, pattern.size(), substitute);
+    pos = data.find(pattern, pos + substitute.size());
+  }
+}
+
+namespace o2
+{
+namespace framework
+{
+
+bool isDataInspectorActive(FairMQDevice& device)
+try {
+  if (device.fConfig == nullptr) {
+    return false;
+  }
+  return device.fConfig->GetProperty<bool>(INSPECTOR_ACTIVATION_PROPERTY);
```

Comment:
This still does a string lookup per message. Can it please be reworked so that is not needed?